Седлал:

DimonDem123 opened the pull request:
Сделал функционал: 
подписаться на пользователя

отписаться от пользователя

получить количество подписчиков пользователя

получить количество подписок пользователя

получить всех подписчиков пользователя

получить все подписки пользователя

DimonDem123 committed:
Седлал:
подписаться на пользователя

отписаться от пользователя

получить количество подписчиков пользователя

получить количество подписок пользователя

получить всех подписчиков пользователя

получить все подписки пользователя

Ana25478 commented on src/main/java/school/faang/user_service/service/user/UserSubscriptionServiceImpl.java:
```diff
@@ -0,0 +1,85 @@
+    private final UserMapper userMapper;
+
+    @Autowired
+    public UserSubscriptionServiceImpl(SubscriptionRepository subscriptionRepository, UserMapper userMapper) {
```
Можно удалить конструктор и повесить на класс аннотацию @RequiredArgsConstructor (создание конструктора для всех финальных полей, и т.к. конструктор будет в таком случае один, Spring будет сразу знать, что пользоваться надо им)

DimonDem123 replied:
Удобно. Не знал

Ana25478 commented on src/main/java/school/faang/user_service/service/user/UserSubscriptionServiceImpl.java:
```diff
@@ -0,0 +1,85 @@
+    public void followUser(long followerId, long followeeId) {
+        if (followerId == followeeId) {
+            log.warn("Пользователь попытался подписаться на самого себя.");
+            return;
```
Тут лучше выбрасывать исключения (почитай в рекоммендациях к выполнению в Jira)

DimonDem123 replied:
Ок

Ana25478 commented on src/main/java/school/faang/user_service/service/user/UserSubscriptionServiceImpl.java:
```diff
@@ -0,0 +1,85 @@
+        boolean alreadyFollower = subscriptionRepository.existsByFollowerIdAndFolloweeId(followerId, followeeId);
+        if (alreadyFollower) {
+            log.warn("Пользователь уже подписан на этого пользователя");
+            return;
```
то же самое - лучше исключение

Ana25478 commented on src/main/java/school/faang/user_service/service/user/UserSubscriptionServiceImpl.java:
```diff
@@ -0,0 +1,85 @@
+            log.warn("Пользователь не может отписаться от самого себя");
+            return;
+        }
+        boolean signed = subscriptionRepository.existsByFollowerIdAndFolloweeId(followerId, followeeId);
```
Мне кажется, не обязателно такие проверки выносить в отдельную локальную переменную, если она используется только раз - я бы ставила сразу в if

DimonDem123 replied:
Логично

Ana25478 reviewed (changes requested):
Оставила пару мелких комментов, вроде всё гуд. Пиши, если есть вопросы

hohlovps95 commented on src/main/java/school/faang/user_service/controller/user/UserSubscriptionController.java:
```diff
@@ -0,0 +1,59 @@
+
+import java.util.List;
+
+@RestController
```
Забыл аннотацию @RequiredArgsConstructor. - конструктор классов
Желательно использовать аннотацию типо такой @RequestMapping("/api/v1")
Это нужно, чтобы разработчики и пользователи понимали с какой версией апи они работают

DimonDem123 replied:
Не знал. Постараюсь запомнить

hohlovps95 commented on src/main/java/school/faang/user_service/controller/user/UserSubscriptionController.java:
```diff
@@ -0,0 +1,59 @@
+        this.userContext = userContext;
+    }
+
+    @PostMapping("/follow")
```
В пути желательно избегать глаголов и прилагательных, желательно использовать существительные

DimonDem123 replied:
Будет сделано

hohlovps95 commented on src/main/java/school/faang/user_service/controller/user/UserSubscriptionController.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    @PostMapping("/unfollow")
+    public void unfollowUser(@RequestParam long followeeId) {
```
Нет проверки на null, я бы использовал аннотации jakarta и валидацию, чтобы не писать дополнительные if

DimonDem123 replied:
Сделано

hohlovps95 commented on src/main/java/school/faang/user_service/service/user/UserSubscriptionServiceImpl.java:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    @Override
+    public List<UserDto> getFollowers(long followeeId) {
```
Я бы добавил проверку, на пустой лист, если лист пустой, выбрасывать ошибку, что ничего не найдено

DimonDem123 replied:
Добавил через анотацию @Valid если можно по другому жду подсказки

hohlovps95 replied:
Чтобы разработчикам было понятно, что именно должно быть во входящем сообщение. Нужно в самом DTO указывать аннотации Jakarta над полями, например, @NotNull

hohlovps95 commented on src/main/java/school/faang/user_service/service/user/UserSubscriptionServiceImpl.java:
```diff
@@ -0,0 +1,85 @@
+
+    @Override
+    public List<UserDto> getFollowees(long followerId) {
+        return subscriptionRepository.findByFollowerId(followerId)
```
Смотри комментарий выше

DimonDem123 replied:
Спасибо за проверку

DimonDem123 committed:
Попытка 2
Добавлено:
1. @RequiredArgsConstructor где нужны
2. Добавлен @RequestMapping для отслеживания версии
3. Переименованы некоторые названия Mapping полей
4. Добавлена проверка на постое значения через jakart

DimonDem123 committed:
Попытка 2
Добавлено:
1. @RequiredArgsConstructor где нужны
2. Добавлен @RequestMapping для отслеживания версии
3. Переименованы некоторые названия Mapping полей
4. Добавлена проверка на постое значения через jakart

DimonDem123 committed:
Update UserSubscriptionServiceImpl.java

hohlovps95 commented on src/main/java/school/faang/user_service/controller/user/UserSubscriptionController.java:
```diff
@@ -0,0 +1,57 @@
+    private final UserSubscriptionService userSubscriptionService;
+    private final UserContext userContext;
+
+    @PostMapping("/Subscription")
```
Я бы путь всё таки указывал с маленькой буквы.

hohlovps95 commented on src/main/java/school/faang/user_service/controller/user/UserSubscriptionController.java:
```diff
@@ -0,0 +1,57 @@
+
+    @PostMapping("/Subscription")
+    public void followUser(@RequestParam long followeeId) {
+        long followerId = userContext.getUserId();
```
Нужна проверка на null. Можно выполнить через Jakarta

hohlovps95 commented on src/main/java/school/faang/user_service/controller/user/UserSubscriptionController.java:
```diff
@@ -0,0 +1,57 @@
+        userSubscriptionService.followUser(followerId, followeeId);
+    }
+
+    @PostMapping("/disclaimer")
```
Лучше избегать глаголов в пути, следует использовать существительные

hohlovps95 commented on src/main/java/school/faang/user_service/service/user/UserSubscriptionServiceImpl.java:
```diff
@@ -0,0 +1,77 @@
+    @Override
+    public void followUser(long followerId, long followeeId) {
+        if (followerId == followeeId) {
+            throw new ForbiddenException("Пользователь попытался подписаться на самого себя.");
```
Лучше добавить в лог ещё id, какой именно пользователь на кого пытался подписаться

hohlovps95 commented on src/main/java/school/faang/user_service/service/user/UserSubscriptionServiceImpl.java:
```diff
@@ -0,0 +1,77 @@
+        }
+
+        if (subscriptionRepository.existsByFollowerIdAndFolloweeId(followerId, followeeId)) {
+            throw new ForbiddenException("Пользователь уже подписан на этого пользователя");
```
Смотри комментарий выше

hohlovps95 commented on src/main/java/school/faang/user_service/service/user/UserSubscriptionServiceImpl.java:
```diff
@@ -0,0 +1,77 @@
+            throw new ForbiddenException("Пользователь не подписывался, чтобы отписаться.");
+        }
+        subscriptionRepository.unfollowUser(followerId, followeeId);
+        log.info("Вы отписались от пользователя");
```
Аналогично, лучше добавлять айди какие пользаки куда пытались подписаться

DimonDem123 replied:
Сделано

Tien-PhanT commented on src/main/java/school/faang/user_service/service/user/UserSubscriptionServiceImpl.java:
```diff
@@ -0,0 +1,77 @@
+
+    @Override
+    public CountResponse getFollowersCount(long followeeId) {
+        long count = subscriptionRepository.findFollowersAmountByFolloweeId(followeeId);
```
Тут можно лог добавить

DimonDem123 replied:
ок

DimonDem123 committed:
Добавил:
логи в UserSubscriptionServiceImpl
проверку на Null с помощью Jacarda

DimonDem123 committed:
Добавил:
логи в UserSubscriptionServiceImpl
проверку на Null с помощью Jacarda
Добавил тесты на UserSubscriptionServiceImpl
